src/ut_control_plane.c

```diff
@@ -0,0 +1,394 @@
+        callback_echo,      // callback function
+        0,                  // per-session data size
+        0,                  // rx buffer size
+        0,                  // id
+        NULL,               // reserved
+        0                   // tx packet size
+    },
+    { NULL, NULL, 0, 0 }    // end of list
+};
+
+ut_controlPlane_instance_t *UT_ControlPlane_Init( uint32_t monitorPort )
+{
+    ut_cp_instance_internal_t *pInstance = calloc(1, sizeof(ut_cp_instance_internal_t));
+
+    if ( pInstance == NULL )
+    {
+        //assert( pInstance != NULL );
+        printf("Malloc was not able to provide memory\n");
+        return NULL;
+    }
+
+    pInstance->magic = UT_CP_MAGIC;
+
+    pInstance->info.port = monitorPort;
+    pInstance->info.iface = NULL;
+    pInstance->info.protocols = protocols;
+
+    pInstance->context = lws_create_context(&pInstance->info);
+    if (!pInstance->context)
+    {
+        printf("Error creating libwebsockets context\n");
+        pInstance->magic = !UT_CP_MAGIC;
+        free( pInstance );
+        return NULL;
+    }
+
+    exit_request = false;
+
+    return (ut_controlPlane_instance_t *)pInstance;
+}
+
+void UT_ControlPlane_Exit( ut_controlPlane_instance_t *pInstance )
+{
+    ut_cp_instance_internal_t *pInternal = validateCPInstance(pInstance);
+
+    if (pInternal == NULL)
+    {
+        return;
+    }
+
+    cp_message_t msg;
+    msg.status = EXIT_REQUESTED;
+    UT_CONTROL_PLANE_DEBUG("EXIT_REQUESTED message from [%s] \n", __func__);
+    enqueue_message(&msg);
+    UT_CONTROL_PLANE_DEBUG("msg count inside CP_exit = %d\n", message_count);
+
+    // Wait for the thread to finish
+    if (pthread_join(pInternal->thread, NULL) != 0)
+    {
+        UT_CONTROL_PLANE_ERROR("Failed to join thread(1st) from instance = %p\n", pInternal);
+        /*TODO: need to confirm if this return is required*/
+        //lws_context_destroy(pInternal->context);
+        //free(pInternal);
+        //return;
+    }
+    UT_CONTROL_PLANE_DEBUG("Thread 1 is also exited from instance : %p\n", pInternal);
+
+    if (pInternal->context != NULL)
+    {
+        lws_context_destroy(pInternal->context);
+    }
+
+    memset(pInternal, 0, sizeof(ut_cp_instance_internal_t));
+
+    free(pInternal);
+}
+
+void UT_ControlPlane_Start( ut_controlPlane_instance_t *pInstance)
+{
+    ut_cp_instance_internal_t *pInternal = validateCPInstance(pInstance);
+
+    if (pInternal == NULL)
+    {
+        return;
+    }
+
+    pthread_create(&pInternal->thread, NULL, thread_function, (void*) pInternal );
+
+}
+
+ut_control_plane_status_t UT_ControlPlane_RegisterCallbackOnMessage(ut_controlPlane_instance_t *pInstance, char *key, ut_control_callback_t callbackFunction)
+{
+    ut_cp_instance_internal_t *pInternal = (ut_cp_instance_internal_t *)pInstance;
+
+    if ( pInternal == NULL )
+    {
+        UT_CONTROL_PLANE_ERROR("Invalid Handle");
+        return UT_CONTROL_PLANE_STATUS_CALLBACK_LIST_INVALID_HANDLE;
```

Comment:
that's an  `UT_CONTROL_PLANE_STATUS_INVALID_HANDLE` not specific to the `CALLBACK_LIST` function